File: providers/coingecko.go
package providers

import (
	"context"
	"strings"
	"sync"
	"time"

	"crypto-tray/internal/httpclient"
)

// CoinGecko implements the Provider interface for the CoinGecko API
type CoinGecko struct {
	httpClient *httpclient.Client

	// Symbol cache
	cacheMu   sync.RWMutex
	symbols   []SymbolInfo
	coinIDMap map[string]SymbolInfo // keyed by coinID for reverse lookup
	cacheTime time.Time
}

// NewCoinGecko creates a new CoinGecko provider
func NewCoinGecko() *CoinGecko {
	return &CoinGecko{
		httpClient: httpclient.New(httpclient.Config{
			BaseURL:      coingeckoBaseURL,
			Timeout:      coingeckoTimeout,
			APIKeyHeader: coingeckoAPIKeyHeader,
		}),
	}
}

func (c *CoinGecko) ID() string               { return "coingecko" }
func (c *CoinGecko) Name() string             { return "CoinGecko" }
func (c *CoinGecko) RequiresAPIKey() bool     { return false }
func (c *CoinGecko) DefaultCoinIDs() []string { return []string{"bitcoin", "ethereum", "solana"} }

func (c *CoinGecko) SetAPIKey(key string) {
	c.httpClient.SetAPIKey(key)
}

// FetchPrices retrieves prices for multiple cryptocurrencies by coinID
func (c *CoinGecko) FetchPrices(ctx context.Context, coinIDs []string) ([]*PriceData, error) {
	if len(coinIDs) == 0 {
		return []*PriceData{}, nil
	}

	query := CoinGeckoPriceQuery{
		IDs:              strings.Join(coinIDs, ","),
		VsCurrencies:     coingeckoCurrency,
		Include24hChange: "true",
	}

	var result CoinGeckoPriceResponse
	if err := c.httpClient.GetWithQuery(ctx, "/simple/price", query, &result); err != nil {
		return nil, err
	}

	prices := make([]*PriceData, 0, len(coinIDs))
	for coinIdx := range coinIDs {
		coinID := coinIDs[coinIdx]
		data, ok := result[coinID]
		if !ok {
			continue
		}
		prices = append(prices, &PriceData{
			CoinID:    coinID,
			Symbol:    c.coinIDToSymbol(coinID),
			Price:     data.USD,
			Change24h: data.USDChange24,
		})
	}

	return prices, nil
}

// FetchSymbols fetches top cryptocurrencies by market cap from CoinGecko API
func (c *CoinGecko) FetchSymbols(ctx context.Context) ([]SymbolInfo, error) {
	// Check cache first
	c.cacheMu.RLock()
	if len(c.symbols) > 0 && time.Since(c.cacheTime) < coingeckoCacheTTL {
		symbols := c.symbols
		c.cacheMu.RUnlock()
		return symbols, nil
	}
	c.cacheMu.RUnlock()

	// Fetch top 250 coins by market cap
	query := CoinGeckoMarketsQuery{
		VsCurrency: coingeckoCurrency,
		Order:      "market_cap_desc",
		PerPage:    "250",
		Page:       "1",
	}

	var coins []CoinGeckoMarketCoin
	if err := c.httpClient.GetWithQuery(ctx, "/coins/markets", query, &coins); err != nil {
		return []SymbolInfo{}, nil
	}

	// Map to SymbolInfo and build lookup map
	symbols := make([]SymbolInfo, 0, len(coins))
	coinIDMap := make(map[string]SymbolInfo, len(coins))

	for coinIdx := range coins {
		coin := coins[coinIdx]
		info := SymbolInfo{
			CoinID: coin.ID,
			Symbol: strings.ToUpper(coin.Symbol),
			Name:   coin.Name,
		}
		symbols = append(symbols, info)
		coinIDMap[coin.ID] = info
	}

	// Update cache
	c.cacheMu.Lock()
	c.symbols = symbols
	c.coinIDMap = coinIDMap
	c.cacheTime = time.Now()
	c.cacheMu.Unlock()

	return symbols, nil
}

// coinIDToSymbol maps a coinID to its display symbol
func (c *CoinGecko) coinIDToSymbol(coinID string) string {
	c.cacheMu.RLock()
	defer c.cacheMu.RUnlock()

	if info, ok := c.coinIDMap[coinID]; ok {
		return info.Symbol
	}

	return strings.ToUpper(coinID)
}

File: src-tauri/src/providers/coingecko.rs
use async_trait::async_trait;
use reqwest::Client;
use serde::Deserialize;
use std::collections::HashMap;
use tokio::sync::RwLock;

use crate::error::Result;
use crate::providers::traits::Provider;
use crate::providers::types::{PriceData, ProviderInfo, SymbolInfo};

const BASE_URL: &str = "https://api.coingecko.com/api/v3";

#[derive(Debug, Deserialize)]
struct MarketCoin {
    id: String,
    symbol: String,
    name: String,
}

#[derive(Debug, Deserialize)]
struct PriceResponse {
    usd: Option<f64>,
    usd_24h_change: Option<f64>,
}

pub struct CoinGeckoProvider {
    client: Client,
    api_key: Option<String>,
    symbol_map: RwLock<HashMap<String, String>>,
}

impl CoinGeckoProvider {
    pub fn new() -> Self {
        Self {
            client: Client::new(),
            api_key: None,
            symbol_map: RwLock::new(HashMap::new()),
        }
    }
}

#[async_trait]
impl Provider for CoinGeckoProvider {
    fn info(&self) -> ProviderInfo {
        ProviderInfo {
            id: "coingecko".to_string(),
            name: "CoinGecko".to_string(),
            requires_api_key: false,
        }
    }

    fn set_api_key(&mut self, key: String) {
        self.api_key = Some(key);
    }

    fn default_coin_ids(&self) -> Vec<String> {
        vec![
            "bitcoin".to_string(),
            "ethereum".to_string(),
            "solana".to_string(),
        ]
    }

    async fn fetch_symbols(&self) -> Result<Vec<SymbolInfo>> {
        let url = format!(
            "{}/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=250&page=1",
            BASE_URL
        );

        let mut request = self.client.get(&url);
        if let Some(ref key) = self.api_key {
            request = request.header("x-cg-demo-api-key", key);
        }

        let response: Vec<MarketCoin> = request.send().await?.json().await?;

        let mut symbol_map = HashMap::with_capacity(response.len());
        let symbols = response
            .into_iter()
            .map(|c| {
                symbol_map.insert(c.id.clone(), c.symbol.to_uppercase());
                SymbolInfo {
                    id: c.id,
                    symbol: c.symbol.to_uppercase(),
                    name: c.name,
                }
            })
            .collect();

        *self.symbol_map.write().await = symbol_map;

        Ok(symbols)
    }

    async fn fetch_prices(&self, symbols: &[String]) -> Result<Vec<PriceData>> {
        // CoinGecko uses coin IDs, not symbols
        // For simplicity, we'll treat symbols as IDs here
        let ids = symbols.join(",");
        let url = format!(
            "{}/simple/price?ids={}&vs_currencies=usd&include_24hr_change=true",
            BASE_URL, ids
        );

        let mut request = self.client.get(&url);
        if let Some(ref key) = self.api_key {
            request = request.header("x-cg-demo-api-key", key);
        }

        let response: HashMap<String, PriceResponse> = request.send().await?.json().await?;

        let symbol_map = self.symbol_map.read().await;

        let prices = response
            .into_iter()
            .map(|(id, data)| {
                let symbol = symbol_map
                    .get(&id)
                    .cloned()
                    .unwrap_or_else(|| id.to_uppercase());

                PriceData {
                    coin_id: id.clone(),
                    symbol,
                    name: String::new(),
                    price: data.usd.unwrap_or(0.0),
                    change_24h: data.usd_24h_change.unwrap_or(0.0),
                    converted_price: 0.0,
                    currency: "usd".to_string(),
                }
            })
            .collect();

        Ok(prices)
    }
}
